File: backend/app/schemas/donor_dashboard.py
"""Pydantic schemas for the admin donor dashboard endpoints."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

# --- Leaderboard ---


class DonorLeaderboardEntry(BaseModel):
    """Single row in the donor leaderboard."""

    user_id: UUID
    first_name: str
    last_name: str
    email: str
    is_active: bool
    total_given: float = Field(default=0)
    events_attended: int = Field(default=0)
    ticket_total: float = Field(default=0)
    donation_total: float = Field(default=0)
    silent_auction_total: float = Field(default=0)
    live_auction_total: float = Field(default=0)
    buy_now_total: float = Field(default=0)


class DonorLeaderboardResponse(BaseModel):
    """Paginated leaderboard result."""

    items: list[DonorLeaderboardEntry]
    total: int
    page: int
    per_page: int
    pages: int


# --- Donor Profile ---


class EventAttendance(BaseModel):
    event_id: UUID
    event_name: str
    event_date: datetime
    npo_id: UUID
    npo_name: str
    checked_in: bool
    total_given_at_event: float = Field(default=0)


class BidRecord(BaseModel):
    bid_id: UUID
    event_id: UUID
    event_name: str
    npo_id: UUID | None = None
    npo_name: str | None = None
    item_id: UUID
    item_title: str
    item_category: str | None = None
    bid_amount: float
    bid_status: str
    bid_type: str
    created_at: datetime


class DonationRecord(BaseModel):
    donation_id: UUID
    event_id: UUID
    event_name: str
    npo_id: UUID | None = None
    npo_name: str | None = None
    amount: float
    source: str
    is_paddle_raise: bool = False
    created_at: datetime


class TicketRecord(BaseModel):
    purchase_id: UUID
    event_id: UUID
    event_name: str
    npo_id: UUID | None = None
    npo_name: str | None = None
    package_name: str
    quantity: int
    total_price: float
    purchased_at: datetime


class OutbidSummary(BaseModel):
    total_outbid_amount: float = Field(default=0)
    items_bid_on: int = Field(default=0)
    items_won: int = Field(default=0)
    items_lost: int = Field(default=0)
    win_rate: float = Field(default=0)


class CategoryInterest(BaseModel):
    category: str
    bid_count: int = Field(default=0)
    total_bid_amount: float = Field(default=0)
    items_won: int = Field(default=0)


class DonorProfileResponse(BaseModel):
    user_id: UUID
    first_name: str
    last_name: str
    email: str
    phone: str | None = None
    is_active: bool
    total_given: float = Field(default=0)
    events_attended: int = Field(default=0)
    event_history: list[EventAttendance] = Field(default_factory=list)
    bid_history: list[BidRecord] = Field(default_factory=list)
    donation_history: list[DonationRecord] = Field(default_factory=list)
    ticket_history: list[TicketRecord] = Field(default_factory=list)
    category_interests: list[CategoryInterest] = Field(default_factory=list)
    outbid_summary: OutbidSummary = Field(default_factory=OutbidSummary)


# --- Outbid Leaders ---


class OutbidLeaderEntry(BaseModel):
    user_id: UUID
    first_name: str
    last_name: str
    total_outbid_amount: float = Field(default=0)
    items_bid_on: int = Field(default=0)
    items_won: int = Field(default=0)
    items_lost: int = Field(default=0)
    win_rate: float = Field(default=0)


class OutbidLeadersResponse(BaseModel):
    items: list[OutbidLeaderEntry]
    total: int
    page: int
    per_page: int
    pages: int


# --- Bid Wars ---


class BidWarItem(BaseModel):
    item_id: UUID
    item_title: str
    bid_count: int
    highest_bid: float
    won: bool


class BidWarEntry(BaseModel):
    user_id: UUID
    first_name: str
    last_name: str
    bid_war_count: int = Field(default=0)
    total_bids_in_wars: int = Field(default=0)
    top_war_items: list[BidWarItem] = Field(default_factory=list)


class BidWarsResponse(BaseModel):
    items: list[BidWarEntry]
    total: int
    page: int
    per_page: int
    pages: int


# --- Category Breakdown ---


class GivingTypeEntry(BaseModel):
    category: str
    total_amount: float = Field(default=0)
    donor_count: int = Field(default=0)


class AuctionCategoryEntry(BaseModel):
    category: str
    total_bid_amount: float = Field(default=0)
    total_revenue: float = Field(default=0)
    bid_count: int = Field(default=0)
    item_count: int = Field(default=0)


class CategoryBreakdownResponse(BaseModel):
    giving_type_breakdown: list[GivingTypeEntry] = Field(default_factory=list)
    auction_category_breakdown: list[AuctionCategoryEntry] = Field(default_factory=list)

File: frontend/fundrbolt-admin/src/services/donor-dashboard.ts
import apiClient from '@/lib/axios'

export interface DonorLabelSummary {
  id: string
  name: string
  color: string | null
  is_system_default?: boolean
  is_suggested?: boolean
  source?: string
}

export interface SurveyAnswerSummary {
  question_text: string
  option_text: string
}

export interface SurveyResponseSummary {
  event_id: string
  event_name: string
  status: 'completed' | 'skipped'
  completed_at: string | null
  discount_cents_applied: number
  answers: SurveyAnswerSummary[]
}

// --- Types ---

export interface DonorLeaderboardEntry {
  user_id: string
  first_name: string
  last_name: string
  email: string
  is_active: boolean
  total_given: number
  events_attended: number
  ticket_total: number
  donation_total: number
  silent_auction_total: number
  live_auction_total: number
  buy_now_total: number
  survey_completed: boolean
  donor_labels: DonorLabelSummary[]
  survey_answers: Record<string, string>
}

export interface DonorLeaderboardResponse {
  items: DonorLeaderboardEntry[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface EventAttendance {
  event_id: string
  event_name: string
  event_date: string
  npo_id: string
  npo_name: string
  checked_in: boolean
  total_given_at_event: number
}

export interface BidRecord {
  bid_id: string
  event_id: string
  event_name: string
  npo_id?: string
  npo_name?: string
  item_id: string
  item_title: string
  item_category: string | null
  bid_amount: number
  bid_status: string
  bid_type: string
  created_at: string
}

export interface DonationRecord {
  donation_id: string
  event_id: string
  event_name: string
  npo_id?: string
  npo_name?: string
  amount: number
  source: string
  is_paddle_raise: boolean
  created_at: string
}

export interface TicketRecord {
  purchase_id: string
  event_id: string
  event_name: string
  npo_id?: string
  npo_name?: string
  package_name: string
  quantity: number
  total_price: number
  purchased_at: string
}

export interface OutbidSummary {
  total_outbid_amount: number
  items_bid_on: number
  items_won: number
  items_lost: number
  win_rate: number
}

export interface CategoryInterest {
  category: string
  bid_count: number
  total_bid_amount: number
  items_won: number
}

export interface DonorProfileResponse {
  user_id: string
  first_name: string
  last_name: string
  email: string
  phone: string | null
  is_active: boolean
  total_given: number
  events_attended: number
  event_history: EventAttendance[]
  bid_history: BidRecord[]
  donation_history: DonationRecord[]
  ticket_history: TicketRecord[]
  category_interests: CategoryInterest[]
  outbid_summary: OutbidSummary
  donor_labels: DonorLabelSummary[]
  survey_completed: boolean
  survey_response: SurveyResponseSummary | null
}

export interface OutbidLeaderEntry {
  user_id: string
  first_name: string
  last_name: string
  total_outbid_amount: number
  items_bid_on: number
  items_won: number
  items_lost: number
  win_rate: number
}

export interface OutbidLeadersResponse {
  items: OutbidLeaderEntry[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface BidWarItem {
  item_id: string
  item_title: string
  bid_count: number
  highest_bid: number
  won: boolean
}

export interface BidWarEntry {
  user_id: string
  first_name: string
  last_name: string
  bid_war_count: number
  total_bids_in_wars: number
  top_war_items: BidWarItem[]
}

export interface BidWarsResponse {
  items: BidWarEntry[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface GivingTypeEntry {
  category: string
  total_amount: number
  donor_count: number
}

export interface AuctionCategoryEntry {
  category: string
  total_bid_amount: number
  total_revenue: number
  bid_count: number
  item_count: number
}

export interface CategoryBreakdownResponse {
  giving_type_breakdown: GivingTypeEntry[]
  auction_category_breakdown: AuctionCategoryEntry[]
}

// --- Query Params ---

export interface LeaderboardParams {
  event_id?: string
  npo_id?: string
  sort_by?: string
  sort_order?: 'asc' | 'desc'
  search?: string
  filter_col?: string
  filter_min?: number
  filter_max?: number
  label_ids?: string[]
  page?: number
  per_page?: number
}

export interface SurveyAnswerColumn {
  id: string
  text: string
}

export interface SurveyAnswerDonorRow {
  user_id: string
  name: string
  answers: Record<string, string>
}

export interface SurveyAnswersResponse {
  questions: SurveyAnswerColumn[]
  donors: SurveyAnswerDonorRow[]
}

export interface EventSurveyAnswerParams {
  eventId: string
  sort_by_question_id?: string
  sort_order?: 'asc' | 'desc'
  filter_question_id?: string
  filter_option_text?: string
}

export interface ScopedParams {
  event_id?: string
  npo_id?: string
}

export interface PaginatedScopedParams extends ScopedParams {
  page?: number
  per_page?: number
}

// --- Service ---

class DonorDashboardService {
  async getLeaderboard(params: LeaderboardParams = {}) {
    const response = await apiClient.get<DonorLeaderboardResponse>(
      '/admin/donor-dashboard/leaderboard',
      { params }
    )
    return response.data
  }

  async getDonorProfile(userId: string, params: ScopedParams = {}) {
    const response = await apiClient.get<DonorProfileResponse>(
      `/admin/donor-dashboard/donors/${userId}`,
      { params }
    )
    return response.data
  }

  async getEventSurveyAnswers(params: EventSurveyAnswerParams) {
    const { eventId, ...queryParams } = params
    const response = await apiClient.get<SurveyAnswersResponse>(
      `/admin/events/${eventId}/survey/donor-answers`,
      { params: queryParams }
    )
    return response.data
  }

  async getOutbidLeaders(params: PaginatedScopedParams = {}) {
    const response = await apiClient.get<OutbidLeadersResponse>(
      '/admin/donor-dashboard/outbid-leaders',
      { params }
    )
    return response.data
  }

  async getBidWars(params: PaginatedScopedParams = {}) {
    const response = await apiClient.get<BidWarsResponse>(
      '/admin/donor-dashboard/bid-wars',
      { params }
    )
    return response.data
  }

  async getCategoryBreakdown(params: ScopedParams = {}) {
    const response = await apiClient.get<CategoryBreakdownResponse>(
      '/admin/donor-dashboard/category-breakdown',
      { params }
    )
    return response.data
  }

  getExportUrl(params: LeaderboardParams = {}): string {
    const searchParams = new URLSearchParams()
    if (params.event_id) searchParams.set('event_id', params.event_id)
    if (params.npo_id) searchParams.set('npo_id', params.npo_id)
    if (params.sort_by) searchParams.set('sort_by', params.sort_by)
    if (params.sort_order) searchParams.set('sort_order', params.sort_order)
    if (params.search) searchParams.set('search', params.search)
    params.label_ids?.forEach((labelId) =>
      searchParams.append('label_ids', labelId)
    )
    const qs = searchParams.toString()
    return `/admin/donor-dashboard/leaderboard/export${qs ? `?${qs}` : ''}`
  }
}

export const donorDashboardService = new DonorDashboardService()
